src/input/json_to_storage.cc

```diff
@@ -440,49 +805,50 @@ StorageBase * JSONToStorage::record_automatic_conversion(JSONPath &p, const Type
 
 
 
-StorageBase * JSONToStorage::make_storage(JSONPath &p, const Type::AbstractRecord *abstr_rec)
+StorageBase * JSONToStorage::make_storage(PathBase &p, const Type::AbstractRecord *abstr_rec)
 {
-    if (p.head()->type() == json_spirit::obj_type) {
-
-        JSONPath type_path(p);
-        if ( type_path.down("TYPE") == NULL ) {
-            if ( ! abstr_rec->get_selection_default().has_value_at_declaration() ) {
-                THROW( ExcInputError() << EI_Specification("Missing key 'TYPE' in AbstractRecord.") << EI_ErrorAddress(p) << EI_InputType(abstr_rec->desc()) );
-            } else { // auto conversion
-            	return abstract_rec_automatic_conversion(p, abstr_rec);
-            }
-        } else {
-            try {
-                // convert to base type to force type dispatch and reference chatching
-                const Type::TypeBase * type_of_type = &( abstr_rec->get_type_selection() );
-                unsigned int descendant_index = (unsigned int)make_storage(type_path, type_of_type )->get_int();
-                return make_storage(p, &( abstr_rec->get_descendant(descendant_index) ) );
-            } catch(Type::Selection::ExcSelectionKeyNotFound &e) {
-
-                THROW( ExcInputError() << EI_Specification("Wrong TYPE='"+Type::EI_KeyName::ref(e)+"' of AbstractRecord.") << EI_ErrorAddress(p) << EI_InputType(abstr_rec->desc()) );
-            }
-        }
-    } else {
-        if ( ! abstr_rec->get_selection_default().has_value_at_declaration() ) {
-            THROW( ExcInputError() << EI_Specification("The value should be 'JSON object', but we found: ")
-                << EI_ErrorAddress(p) << EI_JSON_Type( json_type_names[ p.head()->type() ] ) << EI_InputType(abstr_rec->desc()) );
-        } else { // auto conversion
-        	return abstract_rec_automatic_conversion(p, abstr_rec);
-        }
-    }
+	if ( p.is_record_type() ) {
+
+		string descendant_name = p.get_descendant_name( abstr_rec->get_type_selection() );
+		if ( descendant_name == "" ) {
+			if ( ! abstr_rec->get_selection_default().has_value_at_declaration() ) {
+				THROW( ExcInputError() << EI_Specification("Missing key 'TYPE' in AbstractRecord.") << EI_ErrorAddress(p.as_string()) << EI_InputType(abstr_rec->desc()) );
+			} else { // auto conversion
+				return abstract_rec_automatic_conversion(p, abstr_rec);
+			}
+		} else {
+			try {
+				unsigned int descendant_index = (unsigned int)abstr_rec->get_type_selection().name_to_int( descendant_name );
+				return make_storage(p, &( abstr_rec->get_descendant(descendant_index) ) );
+			} catch (ExcInputError & e) {
```

Comment:
I do not agree with this catch. It was copied here from make_storage for Selection, but
here it would overwrite a correct information coming form make_storage called form the descendant.
I think we can just remove this catch.